contracts/express/ExpressExecutable.sol

```diff
@@ -3,112 +3,249 @@
+        } else {
+            _execute(sourceChain, sourceAddress, payload);
+        }
     }
 
-    /**
-     * @notice This function is shadowed by the proxy and can be called only internally.
-     * @dev Executes the call with token.
-     * @param sourceChain The source blockchain.
-     * @param sourceAddress The source address.
-     * @param payload The payload.
-     * @param tokenSymbol The token symbol.
-     * @param amount The token amount.
-     */
     function executeWithToken(
-        bytes32,
+        bytes32 commandId,
         string calldata sourceChain,
         string calldata sourceAddress,
         bytes calldata payload,
         string calldata tokenSymbol,
         uint256 amount
     ) external {
-        _executeWithToken(sourceChain, sourceAddress, payload, tokenSymbol, amount);
+        {
+            bytes32 payloadHash = keccak256(payload);
+            if (
+                !gateway.validateContractCallAndMint(
+                    commandId,
+                    sourceChain,
+                    sourceAddress,
+                    payloadHash,
+                    tokenSymbol,
+                    amount
+                )
+            ) revert NotApprovedByGateway();
+        }
+        address expressCaller = _popExpressCallerWithToken(
+            commandId,
+            sourceChain,
+            sourceAddress,
+            payload,
+            tokenSymbol,
+            amount
+        );
+        if (expressCaller != address(0)) {
+            {
+                address gatewayToken = gateway.tokenAddresses(tokenSymbol);
+                IERC20(gatewayToken).safeTransfer(expressCaller, amount);
+            }
+            emit ExpressExecutionWithTokenFulfilled(
+                commandId,
+                sourceChain,
+                sourceAddress,
+                payload,
+                tokenSymbol,
+                amount,
+                expressCaller
+            );
+        } else {
+            _executeWithToken(sourceChain, sourceAddress, payload, tokenSymbol, amount);
+        }
     }
 
-    /**
-     * @dev Defines the logic to execute the call.
-     * @param sourceChain The source blockchain.
-     * @param sourceAddress The source address.
-     * @param payload The payload.
-     * This is a virtual function and needs to be implemented in the derived contracts.
-     */
-    function _execute(
+    function expressExecute(
+        bytes32 commandId,
+        string calldata sourceChain,
+        string calldata sourceAddress,
+        bytes calldata payload
+    ) external payable {
+        if (gateway.isCommandExecuted(commandId)) revert AlreadyExecuted();
+
+        address expressCaller = msg.sender;
+
+        _setExpressCaller(commandId, sourceChain, sourceAddress, payload, expressCaller);
+        _execute(sourceChain, sourceAddress, payload);
+        emit ExpressExecuted(commandId, sourceChain, sourceAddress, payload, expressCaller);
```

Comment:
maybe use payload hash for event, and deriving the express executor slot so less bytes are moved around?